### PR TITLE
Unify command building with minimal integration of feature areas.

### PR DESCRIPTION
**Pull request by mikary**
@bricelam please look at the migrations changes, mostly shimmed new code in via SqlBatchBuilder

@ajcvickers please look at update changes, ~~hack in `ReaderModificationCommandBatch.PopulateParameters` is a workaround for invalid parameter names (i.e. `"@@p0"`) fixing would require either adding `ITypeMapper` to the `ModificationCommandBatch` constructor dependencies (non-trivial), or taking `ITypeMapper` as a parameter to `CommandBuilder.CreateDbCommand` which seems inappropriate.~~

@anpete please look at query changes, the query integration was probably the cleanest in this PR and shows more of the general use case I would expect for this kind of command infrastructure.

### CODE REVIEW

File: src/EntityFramework.Relational/Storage/Commands/RelationalCommandBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            Check.NotEmpty(name, nameof(name));
+
+            var parameter = _parameterList.GetOrAdd(name, value, property);
+
+            _stringBuilder.Append(parameter.Name);
+
+            return this;
+        }
+
+        public virtual RelationalCommandBuilder AppendFormat([NotNull] string sqlFragment, [NotNull] params object[] parameters)
```
Comment (bricelam) [inline src/EntityFramework.Relational/Storage/Commands/RelationalCommandBuilder.cs]: The name of this method is misleading. Perhaps something like `AppendWithParameters` would be better...

**Comment by bricelam** [comment]
Changes to Migrations look good.

**Comment by bricelam** [comment]
Remember to file a bug on transaction suppression.

**Comment by mikary** [comment]
@AndriySvyryd since @ajcvickers is out and it's also related to your feature area, could you take a look at this?

File: src/EntityFramework.Relational/Update/ReaderModificationCommandBatch.cs [hunk 2]
```diff
@@ -101,51 +100,49 @@ protected virtual void UpdateCachedCommandText(int commandPosition)
...
         {
-            var parameterName = columnModification.ParameterName;
-            var originalParameterName = columnModification.OriginalParameterName;
-
-            if (parameterName != null
-                || originalParameterName != null)
+            if (columnModification.ParameterName != null)
             {
-                var property = columnModification.Property;
+                var parameterName = columnModification.ParameterName.Substring(1);
```
Comment (AndriySvyryd) [inline src/EntityFramework.Relational/Update/ReaderModificationCommandBatch.cs]: You need to do this because `RelationalCommandBuilder` uses `ParameterNameGenerator`, right?
Can't you just create a `RelationalCommandBuilder` instance in `CreateStoreCommand` that has a noop `ParameterNameGenerator`?

**Comment by ErikEJ** [comment]
Maybe the providers-beware label should be added to this?

File: src/EntityFramework.Relational/Update/ReaderModificationCommandBatch.cs (hunk 2 above)
Comment (mikary) [inline src/EntityFramework.Relational/Update/ReaderModificationCommandBatch.cs]: I'm leaning toward moving the `ParameterNameGenerator` out of the `RelationalCommandBuilder` since there isn't any other provider-specific / SQL generation logic in the builder.

Comment (AndriySvyryd) [inline src/EntityFramework.Relational/Update/ReaderModificationCommandBatch.cs]: With that the Update changes will look good

**Comment by mikary** [comment]
PR updated based on comments / feedback

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 3]
```diff
@@ -24,38 +25,44 @@ namespace Microsoft.Data.Entity.Query.Sql
...
 
         public virtual SelectExpression SelectExpression => _selectExpression;
 
-        public virtual string GenerateSql(IDictionary<string, object> parameterValues)
+        protected virtual IParameterNameGeneratorFactory ParameterNameGeneratorFactory
+            => _parameterNameGeneratorFactory;
+
+        protected virtual ParameterNameGenerator ParameterNameGenerator => _parameterNameGenerator;
+
+        public virtual RelationalCommand GenerateSql([NotNull] IDictionary<string, object> parameterValues)
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: This should probably be renamed

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 4]
```diff
@@ -67,12 +74,10 @@ public virtual string GenerateSql(IDictionary<string, object> parameterValues)
                 .Create(_selectExpression.GetProjectionTypes().ToArray(), indexMap: null);
         }
 
-        public virtual IReadOnlyList<CommandParameter> Parameters => _commandParameters;
-
-        protected virtual IndentedStringBuilder Sql => _sql;
+        protected virtual RelationalCommandBuilder Sql => _sql;
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: rename

File: src/EntityFramework.Relational/Query/Sql/ISqlQueryGenerator.cs [hunk 5]
```diff
@@ -5,14 +5,13 @@
 using System.Data.Common;
 using JetBrains.Annotations;
 using Microsoft.Data.Entity.Storage;
+using Microsoft.Data.Entity.Storage.Commands;
 
 namespace Microsoft.Data.Entity.Query.Sql
 {
     public interface ISqlQueryGenerator
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/ISqlQueryGenerator.cs]: rename

File: src/EntityFramework.Relational/Query/Sql/ISqlQueryGenerator.cs [hunk 6]
```diff
@@ -5,14 +5,13 @@
...
     {
-        string GenerateSql([NotNull] IDictionary<string, object> parameterValues);
-
-        IReadOnlyList<CommandParameter> Parameters { get; }
+        RelationalCommand GenerateSql([NotNull] IDictionary<string, object> parameterValues);
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/Sql/ISqlQueryGenerator.cs]: rename

File: src/EntityFramework.Relational/Storage/Commands/RelationalCommandBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+{
+    public class RelationalCommandBuilder
+    {
+        private readonly IndentedStringBuilder _stringBuilder = new IndentedStringBuilder();
+        private readonly RelationalParameterList _parameterList = new RelationalParameterList();
+
+        public virtual RelationalCommandBuilder AppendLine()
+            => AppendLine(string.Empty);
+
+        public virtual RelationalCommandBuilder Append([NotNull]object o)
```
Comment (maumar) [inline src/EntityFramework.Relational/Storage/Commands/RelationalCommandBuilder.cs]: formatting, add space after [NotNul]

File: src/EntityFramework.Commands/Migrations/CSharpMigrationOperationGenerator.cs [hunk 8]
```diff
@@ -1133,20 +1133,8 @@ protected virtual void Generate([NotNull] SqlOperation operation, [NotNull] Inde
...
                 builder.Append(_code.Literal(operation.Sql));
 
-                if (operation.SuppressTransaction)
-                {
-                    builder
-                        .AppendLine(",")
-                        .Append("suppressTransaction: true");
-                }
-
                 builder.Append(")");
```
Comment (bricelam) [inline src/EntityFramework.Commands/Migrations/CSharpMigrationOperationGenerator.cs]: Could simplify all of this a bit further to just:
```C#
builder
    .Append(".Sql(")
    .Append(_code.Literal(operation.Sql))
    .Append(")");
```

File: test/EntityFramework.Sqlite.Tests/Storage/SqliteTypeMappingTest.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Data;
+using System.Data.Common;
+using Microsoft.Data.Sqlite;
+
+namespace Microsoft.Data.Entity.Storage
+{
+    public class SqliteTypeMappingTest : RelationalTypeMappingTest
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.Tests/Storage/SqliteTypeMappingTest.cs]: Thanks, these [should have been re-enabled a while ago](https://github.com/aspnet/EntityFramework/pull/2277#discussion_r31185027).

File: src/EntityFramework.Relational/Infrastructure/RelationalSqlExecutor.cs [hunk 10]
```diff
@@ -1,24 +1,33 @@
...
 using JetBrains.Annotations;
-using Microsoft.Data.Entity.Query;
+using Microsoft.Data.Entity.Sql;
 using Microsoft.Data.Entity.Storage;
+using Microsoft.Data.Entity.Storage.Commands;
+using Microsoft.Data.Entity.Utilities;
 
 namespace Microsoft.Data.Entity.Infrastructure
 {
     public class RelationalSqlExecutor
```
Comment (bricelam) [inline src/EntityFramework.Relational/Infrastructure/RelationalSqlExecutor.cs]: What value does this add over `SqlStatementExecutor`?

File: src/EntityFramework.Relational/Storage/ISqlStatementExecutor.cs [hunk 11]
```diff
@@ -6,43 +6,46 @@
 using System.Threading;
 using System.Threading.Tasks;
 using JetBrains.Annotations;
-using Microsoft.Data.Entity.Infrastructure;
+using Microsoft.Data.Entity.Storage.Commands;
 
 namespace Microsoft.Data.Entity.Storage
 {
     public interface ISqlStatementExecutor
```
Comment (bricelam) [inline src/EntityFramework.Relational/Storage/ISqlStatementExecutor.cs]: So much cleaner!

**Comment by bricelam** [comment]
Still looks good from a Migrations point of view.

File: src/EntityFramework.Relational/Infrastructure/RelationalSqlExecutor.cs (hunk 10 above)
Comment (mikary) [inline src/EntityFramework.Relational/Infrastructure/RelationalSqlExecutor.cs]: This class provides an entry point for the `ExecuteSqlCommand` extension and it basically serves as a place to inject the `IRelationalConnection`. The `object[]` overload could move into the `SqlStatementExecutor`, but I think we need to leave the connection out since there are cases where we use different connections in migrations (you'd probably know more about that than I do).

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs [hunk 12]
```diff
@@ -191,5 +183,26 @@ public SqlStatementExecutor([NotNull] ILoggerFactory loggerFactory)
                 connection.Close();
             }
         }
+
+        protected virtual async Task<T> ExecuteAsync<T>(
+            IRelationalConnection connection,
+            Func<Task<T>> action,
```
Comment (anpete) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: Pass connection to Func instead of capturing in the closure.

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs [hunk 13]
```diff
@@ -147,30 +138,31 @@ public SqlStatementExecutor([NotNull] ILoggerFactory loggerFactory)
...
-            IRelationalConnection connection,
-            DbTransaction transaction,
-            string sql)
+        public virtual Task<DbDataReader> ExecuteReaderAsync(
+            [NotNull] IRelationalConnection connection,
+            [NotNull] string sql,
+            CancellationToken cancellationToken = default(CancellationToken))
         {
             Check.NotNull(connection, nameof(connection));
             Check.NotNull(sql, nameof(sql));
```
Comment (anpete) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: NotEmpty?

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs [hunk 14]
```diff
@@ -147,30 +138,31 @@ public SqlStatementExecutor([NotNull] ILoggerFactory loggerFactory)
...
 
-            return Execute(
+            return ExecuteAsync(
                 connection,
                 () =>
-                {
-                    var command = new SqlBatch(sql).CreateCommand(connection, transaction);
-                    Logger.LogCommand(command);
+                    {
+                        var command = new RelationalCommand(sql).CreateDbCommand(connection, _typeMapper);
```
Comment (anpete) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: Missing Dispose?

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs [hunk 15]
```diff
@@ -147,30 +138,31 @@ public SqlStatementExecutor([NotNull] ILoggerFactory loggerFactory)
                 connection,
                 () =>
                     {
-                        var command = new SqlBatch(sql).CreateCommand(connection, transaction);
+                        var command = new RelationalCommand(sql).CreateDbCommand(connection, _typeMapper);
```
Comment (anpete) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: Missing Dispose?

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs [hunk 16]
```diff
@@ -17,128 +17,119 @@ namespace Microsoft.Data.Entity.Storage
...
-                        foreach (var sqlBatch in sqlBatches)
-                        {
-                            var command = sqlBatch.CreateCommand(connection, transaction);
-                            Logger.LogCommand(command);
-
-                            command.ExecuteNonQuery();
-                        }
-                        return default(object);
-                    });
+                        var command = new RelationalCommand(sql).CreateDbCommand(connection, _typeMapper);
```
Comment (anpete) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: Missing Dispose?

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs (hunk 12 above)
Comment (anpete) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: Perhaps move DbCommand creation here so you can centralize and dispose in one place?

File: src/EntityFramework.Relational/Storage/ISqlStatementExecutor.cs [hunk 17]
```diff
@@ -6,43 +6,46 @@
...
     {
-        Task ExecuteNonQueryAsync(
+        void ExecuteNonQuery(
```
Comment (anpete) [inline src/EntityFramework.Relational/Storage/ISqlStatementExecutor.cs]: Unify to a single version of each method. params?

File: src/EntityFramework.Relational/Storage/Commands/RelationalParameterList.cs [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+namespace Microsoft.Data.Entity.Storage.Commands
+{
+    public class RelationalParameterList
+    {
+        private readonly List<RelationalParameter> _parameters = new List<RelationalParameter>();
+
+        public virtual IReadOnlyList<RelationalParameter> RelationalParameters => _parameters;
+
+        public virtual RelationalParameter GetOrAdd(
+            [NotNull] string name,
```
Comment (anpete) [inline src/EntityFramework.Relational/Storage/Commands/RelationalParameterList.cs]: NotEmpty

File: src/EntityFramework.Relational/Query/Sql/RawSqlQueryGenerator.cs [hunk 19]
```diff
@@ -27,47 +29,50 @@ public class RawSqlQueryGenerator : ISqlQueryGenerator
...
             Check.NotNull(sql, nameof(sql));
             Check.NotNull(parameters, nameof(parameters));
+            Check.NotNull(typeMapper, nameof(typeMapper));
 
             _selectExpression = selectExpression;
             _sql = sql;
             _inputParameters = parameters;
-            _commandParameters = new List<CommandParameter>();
             TypeMapper = typeMapper;
+            _parameterNameGeneratorFactory = new ParameterNameGeneratorFactory();
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/Sql/RawSqlQueryGenerator.cs]: Is this pending DI?

Comment (mikary) [inline src/EntityFramework.Relational/Query/Sql/RawSqlQueryGenerator.cs]: Yes, should be trivial to add to the query DI change

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 20]
```diff
@@ -24,38 +25,44 @@ namespace Microsoft.Data.Entity.Query.Sql
...
             [NotNull] SelectExpression selectExpression,
-            [CanBeNull] IRelationalTypeMapper typeMapper)
+            [NotNull] IRelationalTypeMapper typeMapper)
         {
             Check.NotNull(selectExpression, nameof(selectExpression));
+            Check.NotNull(typeMapper, nameof(typeMapper));
 
             _selectExpression = selectExpression;
             TypeMapper = typeMapper;
+            _parameterNameGeneratorFactory = new ParameterNameGeneratorFactory();
```
Comment (AndriySvyryd) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Why is this not obtained from DI?

**Comment by AndriySvyryd** [comment]
Update changes look good

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 20 above)
Comment (mikary) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Getting an item from DI in this part of query is difficult and I'm working on a large change to query that will make fixing this trivial, so this is just an intermediate solution.

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs (hunk 3 above)
Comment (mikary) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: I agree, but I think that is beyond the scope of this PR. Once we have the basic unified commands through, I'd like to do separate PRs for each of the feature areas to cleanup/refactor around the changes.

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs (hunk 12 above)
Comment (mikary) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: Are you thinking we would move the `DbCommand` and `DbParameter` creation logic out of `RelationalCommand` and `RelationalParameter` and have everything depend on the `SqlStatementExecutor`?

That sounds like it could be a good factoring, but I would suggest we save that kind of change for a later PR so that the individual feature areas can be adjusted to use the executor first as there are still some areas that are not very DI friendly.

File: src/EntityFramework.Relational/Infrastructure/RelationalSqlExecutor.cs (hunk 10 above)
Comment (bricelam) [inline src/EntityFramework.Relational/Infrastructure/RelationalSqlExecutor.cs]: SQL Server needs to execute `CREATE DATABASE` statements against `master`, but that's the only special case I can think of. We should look at simplifying these in another wave. There are a few components in Migrations that depend on `IRelationalConnection` just to pass it into the executor. That can probably be simplified.

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs (hunk 12 above)
Comment (anpete) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: No, was just calling out a way to DRY up the ExecuteX methods. Specifically, the call to CreateDbCommand and the missing using.

Comment (mikary) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: Ah, yes, that does improve the factoring quite a bit. :)

File: src/EntityFramework.Relational/Storage/ISqlStatementExecutor.cs (hunk 17 above)
Comment (mikary) [inline src/EntityFramework.Relational/Storage/ISqlStatementExecutor.cs]: I don't think the params pattern helps here, it requires a lot of existing code to add `ToArray()` calls and makes consistency with the async version weird:
```cs
ExecuteQueryAsync(
    [NotNull] IRelationalConnection connection,
    [NotNull] CancellationToken cancellationToken,
    [NotNull] params RelationalCommand[] relationalCommands)
```
this is also more of a provider surface than public surface, so we might want to revisit in API review but this doesn't seem like a pri0 decsision

File: src/EntityFramework.Relational/EntityFramework.Relational.csproj [hunk 21]
```diff
@@ -264,15 +267,15 @@
     <Compile Include="Update\IBatchExecutor.cs" />
     <Compile Include="Update\ICommandBatchPreparer.cs" />
     <Compile Include="Update\IModificationCommandBatchFactory.cs" />
-    <Compile Include="Update\IParameterNameGeneratorFactory.cs" />
+    <Compile Include="Sql\IParameterNameGeneratorFactory.cs" />
```
Comment (bricelam) [inline src/EntityFramework.Relational/EntityFramework.Relational.csproj]: Do we need to introduce a new namespace? Should this go under `Storage` instead?

File: src/EntityFramework.Relational/Query/Expressions/SelectExpression.cs [hunk 22]
```diff
@@ -713,9 +713,6 @@ protected override Expression VisitChildren(ExpressionVisitor visitor)
             return this;
         }
 
-        public override string ToString() 
-            => new DefaultQuerySqlGenerator(this, null).GenerateSql(new Dictionary<string, object>());
-
```
Comment (bricelam) [inline src/EntityFramework.Relational/Query/Expressions/SelectExpression.cs]: Is this removal intentional?

File: src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 23]
```diff
@@ -24,38 +25,44 @@ namespace Microsoft.Data.Entity.Query.Sql
     public class DefaultQuerySqlGenerator : ThrowingExpressionVisitor, ISqlExpressionVisitor, ISqlQueryGenerator
     {
         private readonly SelectExpression _selectExpression;
+        private readonly IParameterNameGeneratorFactory _parameterNameGeneratorFactory;
```
Comment (bricelam) [inline src/EntityFramework.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: No need for a field; just use a read-only auto-property.

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs [hunk 24]
```diff
@@ -17,174 +17,169 @@ namespace Microsoft.Data.Entity.Storage
...
-                cancellationToken);
+            foreach (var command in relationalCommands)
+            {
+                Execute<object>(
+                    connection,
+                    command,
+                    c => c.ExecuteNonQuery());
+            }
+
+            connection.Close();
```
Comment (bricelam) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: Wrap inside `finally` so reference counting doesn't get of on exceptions.

File: src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs [hunk 25]
```diff
@@ -17,174 +17,169 @@ namespace Microsoft.Data.Entity.Storage
...
-            {
-                connection.Close();
+                await ExecuteAsync(
+                    connection,
+                    command,
+                    async c => await c.ExecuteNonQueryAsync(cancellationToken),
+                    cancellationToken);
             }
+
+            connection.Close();
```
Comment (bricelam) [inline src/EntityFramework.Relational/Storage/SqlStatementExecutor.cs]: `finally`

**Comment by bricelam** [comment]
:shipit: After addressing [new feedback](https://github.com/aspnet/EntityFramework/pull/2758#discussion-diff-37682496), of course.

File: src/EntityFramework.Relational/Query/Expressions/SelectExpression.cs (hunk 22 above)
Comment (mikary) [inline src/EntityFramework.Relational/Query/Expressions/SelectExpression.cs]: Yep, this is just a tweak for the `RelationalExpressionPrinter` related to some of the `IRelationalTypeMapper` changes in `DefaultQuerySqlGenerator`
